pkg/services/ngalert/state/manager.go

```diff
@@ -0,0 +1,118 @@
+package state
+
+import (
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	ngModels "github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+type Manager struct {
+	cache *cache
+	quit  chan struct{}
+	Log   log.Logger
+}
+
+func NewManager(logger log.Logger) *Manager {
+	manager := &Manager{
+		cache: newCache(),
+		quit:  make(chan struct{}),
+		Log:   logger,
+	}
+	go manager.cleanUp()
+	return manager
+}
+
+func (st *Manager) Close() {
+	st.quit <- struct{}{}
+}
+
+func (st *Manager) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result) *State {
+	return st.cache.getOrCreate(alertRule, result)
+}
+
+func (st *Manager) set(entry *State) {
+	st.cache.set(entry)
+}
+
+func (st *Manager) Get(id string) (*State, error) {
+	return st.cache.get(id)
+}
+
+//Used to ensure a clean cache on startup
+func (st *Manager) ResetCache() {
+	st.cache.reset()
+}
+
+func (st *Manager) ProcessEvalResults(alertRule *ngModels.AlertRule, results eval.Results) []*State {
+	st.Log.Info("state tracker processing evaluation results", "uid", alertRule.UID, "resultCount", len(results))
+	var states []*State
+	for _, result := range results {
+		s := st.setNextState(alertRule, result)
+		states = append(states, s)
+	}
+	st.Log.Debug("returning changed states to scheduler", "count", len(states))
+	return states
+}
+
+//TODO: When calculating if an alert should not be firing anymore, we should take three things into account:
+// 1. The re-send the delay if any, we don't want to send every firing alert every time, we should have a fixed delay across all alerts to avoid saturating the notification system
```

Comment:
```suggestion
// 1. The re-send delay if any, we don't want to send every firing alert every time, we should have a fixed delay across all alerts to avoid saturating the notification system
```